planner/core/rule_topn_push_down.go

```diff
@@ -62,57 +64,67 @@ func (lt *LogicalTopN) setChild(p LogicalPlan) LogicalPlan {
 			limitHints: lt.limitHints,
 		}.Init(lt.ctx, lt.blockOffset)
 		limit.SetChildren(p)
+		appendTopNPushDownTraceStep(limit, p, opt)
 		return limit
 	}
 	// Then lt must be topN.
 	lt.SetChildren(p)
+	appendTopNPushDownTraceStep(lt, p, opt)
 	return lt
 }
 
-func (ls *LogicalSort) pushDownTopN(topN *LogicalTopN) LogicalPlan {
+func (ls *LogicalSort) pushDownTopN(topN *LogicalTopN, opt *logicalOptimizeOp) LogicalPlan {
 	if topN == nil {
-		return ls.baseLogicalPlan.pushDownTopN(nil)
+		return ls.baseLogicalPlan.pushDownTopN(nil, opt)
 	} else if topN.isLimit() {
 		topN.ByItems = ls.ByItems
-		return ls.children[0].pushDownTopN(topN)
+		appendSortPassByItemsTraceStep(ls, topN, opt)
+		return ls.children[0].pushDownTopN(topN, opt)
 	}
 	// If a TopN is pushed down, this sort is useless.
-	return ls.children[0].pushDownTopN(topN)
+	return ls.children[0].pushDownTopN(topN, opt)
 }
 
-func (p *LogicalLimit) convertToTopN() *LogicalTopN {
-	return LogicalTopN{Offset: p.Offset, Count: p.Count, limitHints: p.limitHints}.Init(p.ctx, p.blockOffset)
+func (p *LogicalLimit) convertToTopN(opt *logicalOptimizeOp) *LogicalTopN {
+	topn := LogicalTopN{Offset: p.Offset, Count: p.Count, limitHints: p.limitHints}.Init(p.ctx, p.blockOffset)
+	opt.appendStepToCurrent(topn.ID(), topn.TP(), "Limit can be converted into TopN", fmt.Sprintf("%v[%v] is converted into %v[%v]",
```

Comment:
through the `%v[%v]` format looks more structured to describe operators, it's more important to keep consistent with the format `%v_%v` used by the result of `explain`.

One of the typical scenario of using optimizer tracing is:
- execute `explain` to see the execution plan of a slow query
- execute the optimizer tracing to see the internals of why the execution plan is generated, how each operator in the plan is generated

Reply:
updated